awx/main/models/unified_jobs.py

```diff
@@ -338,6 +339,23 @@ def notification_templates(self):
         from awx.main.models.notifications import NotificationTemplate
         return NotificationTemplate.objects.none()
 
+    def resolve_execution_environment(self):
+        """
+        Return the execution environment that should be used when creating a new job.
+        """
+        if self.execution_environment is not None:
+            return self.execution_environment
+        if getattr(self, 'project_id', None) and self.project.default_environment is not None:
+            return self.project.default_environment
+        if getattr(self, 'organization', None) and self.organization.default_environment is not None:
+            return self.organization.default_environment
+        if getattr(self, 'inventory', None) and self.inventory.organization is not None:
+            if self.inventory.organization.default_environment is not None:
+                return self.inventory.organization.default_environment
```

Comment:
It seems strange that inventory would come after organization in the selection logic.

Also, that creates a lot of confusion with inventory updates, since there would be an EE on inventory and an EE on the inventory source. IIRC, with custom virtual environments, inventory sources could have a custom venv specified, but inventory could not.

Reply:
`Inventory` does not have an EE.  It is used here solely for the purpose of inferring an `Organization`, for jobs that fail to have one more directly defined.

`InventorySource` _does_ have an EE, however, I have been very clear about maintaining a strict separation between "EE as explicitly set for use when executing" (always `self.execution_environment`) and "EE as a fallback option" (always `self.default_environment`).

Reply:
To be more clear, `InventorySource` has `self.execution_environment`.  The only models that have `self.default_environment` are / will be `Organization` and `Project`.

Reply:
Thanks, we're on the same page about inventory source.

But in what case is it desirable to use `inventory.organization.default_environment`? I don't think that's desirable to JTs, it would confuse more than it would help. I could see that for the case of inventory updates where the inventory source does not set an EE.

So maybe make this conditional specific to the inventory update job type?

Reply:
Is there a non-contrived case where a JT would not have an Org but would wind up getting its EE from an inventory Org instead?  And would using that EE actually be harmful instead of falling back all of the way to the global default EE?